### PR TITLE
.Net - Agents Fix: Capture Function-Calling Content (ChatCompletionAgent)

### PR DESCRIPTION
**Pull request by crickman**
### Motivation and Context

Result from function-calling not being captured in agent-chat history for ChatCompletionAgent.

### Description

Updated logic to:

1. Capture mutated chat-history, in addition to the iterated messages.
2. Suppress exposing function-content as `AgentChat` interaction (while retaining in history).
3. "Tool" role transformed to "Assistant" for OpenAI Assistant API

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [X] All unit tests pass, and I have added new tests where possible
- [X] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Agents/Core/ChatCompletionAgent.cs [hunk 1]
```diff
@@ -34,14 +34,25 @@ public sealed class ChatCompletionAgent : ChatHistoryKernelAgent
...
+
+        IReadOnlyList<ChatMessageContent> messages =
             await chatCompletionService.GetChatMessageContentsAsync(
                 chat,
                 this.ExecutionSettings,
                 this.Kernel,
                 cancellationToken).ConfigureAwait(false);
 
-        foreach (var message in messages ?? [])
+        for (int messageIndex = messageCount; messageIndex < chat.Count; messageIndex++)
```
Comment (SergeyMenshykh) [inline dotnet/src/Agents/Core/ChatCompletionAgent.cs:46]: nit: messageIndex < messageCount

File: dotnet/src/Agents/Core/ChatCompletionAgent.cs [hunk 2]
```diff
@@ -34,14 +34,25 @@ public sealed class ChatCompletionAgent : ChatHistoryKernelAgent
...
+        {
+            ChatMessageContent message = chat[messageIndex];
+
+            message.AuthorName = this.Name;
+
+            yield return message;
+        }
```
Comment (SergeyMenshykh) [inline dotnet/src/Agents/Core/ChatCompletionAgent.cs:53]: Won't returning the entire chat history by this method lead to duplicates in the chat history stored in the ChatHistoryChannel class?
image

File: dotnet/src/Agents/Abstractions/AgentChat.cs [hunk 3]
```diff
@@ -179,6 +179,12 @@ await foreach (var message in channel.InvokeAsync(agent, cancellationToken).Conf
                 this.History.Add(message);
                 messages.Add(message);
 
+                if (message.Role == AuthorRole.Tool || message.Items.All(i => i is FunctionCallContent))
```
Comment (SergeyMenshykh) [inline dotnet/src/Agents/Abstractions/AgentChat.cs:182]: Is there a reason to filter out function call content and results? What I've noticed is that if both are excluded from a chat history and the chat history is sent to LLM again, the scenario that is possible when the whole chat is modeled as an agent, the LLM will call the functions again.

File: dotnet/src/Agents/Core/ChatCompletionAgent.cs (hunk 2 above)
Comment (crickman) [reply dotnet/src/Agents/Core/ChatCompletionAgent.cs:53]: I believe only the incremental messages are added.

File: dotnet/src/Agents/Abstractions/AgentChat.cs (hunk 3 above)
Comment (crickman) [reply dotnet/src/Agents/Abstractions/AgentChat.cs:182]: This is only filtering out what is returned to the user.  The content is included in the history.

This is intended to align with core SK behavior

File: dotnet/src/Agents/Core/ChatCompletionAgent.cs (hunk 1 above)
Comment (crickman) [reply dotnet/src/Agents/Core/ChatCompletionAgent.cs]: Updated

Comment (SergeyMenshykh) [reply dotnet/src/Agents/Core/ChatCompletionAgent.cs]: That's a wrong advise. Please rollback it if it's applied.

File: dotnet/src/Agents/Core/ChatCompletionAgent.cs (hunk 2 above)
Comment (SergeyMenshykh) [reply dotnet/src/Agents/Core/ChatCompletionAgent.cs:53]: Right

File: dotnet/src/Agents/Core/ChatCompletionAgent.cs (hunk 1 above)
Comment (crickman) [reply dotnet/src/Agents/Core/ChatCompletionAgent.cs]: Oh right...good catch.

Comment (crickman) [reply dotnet/src/Agents/Core/ChatCompletionAgent.cs:46]: Rolled-back.

**Review by SergeyMenshykh — APPROVED** [review]

**Review by alliscode — APPROVED** [review]

File: dotnet/src/Agents/Abstractions/AgentChat.cs [hunk 4]
```diff
@@ -179,6 +179,12 @@ await foreach (var message in channel.InvokeAsync(agent, cancellationToken).Conf
...
+                {
+                    // Don't explore internal messages to caller.
```
Comment (markwallace-microsoft) [inline dotnet/src/Agents/Abstractions/AgentChat.cs]: Should it be 'Don't expose' ?

**Pull request merged**